shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -556,24 +556,51 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
   def computeVariances(tyDefs: List[TypeDef], ctx: Ctx): Unit = {
     var varianceUpdated: Bool = false;
 
-    def fieldVarianceHelper(fieldTy: FieldType, curVariance: VarianceInfo)(implicit tyDef: TypeDef): Unit = {
+    def fieldVarianceHelper(fieldTy: FieldType, curVariance: VarianceInfo)(implicit tyDef: TypeDef, visited: MutSet[TypeVariable]): Unit = {
         fieldTy.lb.foreach(lb => updateVariance(lb, curVariance.flip))
         updateVariance(fieldTy.ub, curVariance)
     }
 
-    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit tyDef: TypeDef): Unit =
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit tyDef: TypeDef, visited: MutSet[TypeVariable]): Unit =
```

Comment:
As I mentioned in my original comment, this should be a `MutSet[(Bool, TV)]`. Can please you make sure to follow my suggestions so I don't have to repeat myself?
Finding a variable `tv` in positive position should set `visited += true -> tv`. Negative position `visited += false -> tv`. Invariant position, both.